File: app/src/main/java/io/diaryofrifat/code/rifbase/ui/base/BaseViewModel.java
package io.diaryofrifat.code.rifbase.ui.base;

import android.arch.lifecycle.ViewModel;

public class BaseViewModel<V extends MvpView, P extends Presenter<V>> extends ViewModel {
    private P presenter;

    P getPresenter() {
        return this.presenter;
    }

    void setPresenter(P presenter) {
        if (this.presenter == null) {
            this.presenter = presenter;
        }
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        presenter.onPresenterDestroy();
        presenter = null;
    }
}

File: app/src/main/java/io/diaryofrifat/code/basemvpkt/ui/base/component/BaseViewModel.kt
package io.diaryofrifat.code.basemvpkt.ui.base.component

import androidx.lifecycle.ViewModel
import io.diaryofrifat.code.basemvpkt.ui.base.callback.MvpView
import io.diaryofrifat.code.basemvpkt.ui.base.callback.Presenter

class BaseViewModel<V : MvpView, P : Presenter<V>> : ViewModel() {
    /**
     * Fields
     * */
    private var presenter: P? = null

    /**
     * Getter and setter methods for presenter
     * */
    internal fun getPresenter(): P? {
        return this.presenter
    }

    internal fun setPresenter(presenter: P) {
        if (this.presenter == null) {
            this.presenter = presenter
        }
    }

    /**
     * This method is fired when to clear the presenter
     * */
    override fun onCleared() {
        super.onCleared()
        presenter?.onPresenterDestroy()
        presenter = null
    }
}
